test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/JSOptionalTest.scala

```diff
@@ -0,0 +1,291 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js Test Suite        **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2016, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+package org.scalajs.testsuite.jsinterop
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation._
+
+import org.junit.Assert._
+import org.junit.Test
+
+import org.scalajs.testsuite.utils.JSAssert._
+
+class JSOptionalTest {
+  import JSOptionalTest._
+
+  @Test def classImplementsTraitWithOptional(): Unit = {
+    val obj = new ClassImplementsTraitWithOptional
+
+    assertEquals(js.undefined, obj.x)
+    assertFalse(obj.hasOwnProperty("x"))
+
+    assertEquals(js.undefined, obj.y)
+    assertFalse(js.Object.hasProperty(obj, "y"))
+
+    assertEquals(js.undefined, obj.y2)
+    assertFalse(js.Object.hasProperty(obj, "y2"))
+
+    assertEquals(js.undefined, obj.z)
+    assertFalse(obj.hasOwnProperty("z"))
+    obj.z = Some(3)
```

Comment:
Oh. That is a very good observation. But in fact the problem is much larger than with optional fields. JS types have JS semantics, so any JS code can add any arbitrary field or method to our JS classes. That means all private fields in JS classes should be protected from being renamed. Ideally they would have `symbol` names, but that wouldn't work in ES5 and would probably have performance issues in ES6. We'll need to think about this, but the bug already exists without this PR so it is a separate issue.

Filed as #2680.